src/misc/msccl/msccl_status.cc

```diff
@@ -27,18 +27,20 @@ static mutex rankStatesMutex;
 static unordered_map<int, shared_ptr<mscclRankState>> rankStates;
 
 static inline mscclRankState& mscclGetRankState(int rank) {
-  static thread_local shared_ptr<mscclRankState> threadRankState = make_shared<mscclRankState>();
-
+  // In the unlikely case of negative rank, return a per-thread state
   if (rank < 0) {
+    static thread_local shared_ptr<mscclRankState> threadRankState(new mscclRankState());
     return *threadRankState;
   }
 
   lock_guard<mutex> lock(rankStatesMutex);
 
   auto rankStateIt = rankStates.find(rank);
   if (rankStateIt == rankStates.end()) {
-    rankStateIt = rankStates.insert(make_pair(rank, make_shared<mscclRankState>(*threadRankState))).first;
-    rankStateIt->second->rank = rank;
+    // Create a per rank threadRankState rather than per thread
+    shared_ptr<mscclRankState> newthreadRankState(new mscclRankState());
+    newthreadRankState->rank = rank;
+    rankStateIt = rankStates.insert(make_pair(rank, newthreadRankState)).first;
```

Comment:
The original code already made a per-rank state rather than a per-thread state. It was copying from the existing thread state in case changes had been made to it earlier in the code path. This is the whole reason a negative rank is not unlikely above. It is by design.

Reply:
It created a per-rank state, but the previous `static thread_local` indicates that if multiple ranks are on the same thread, they will share the state pointer. That's my understanding of `static thread_local`.

Reply:
They will, but the original code was making a copy to preserve any state information that was set before the rank ID was known.